### PR TITLE
SAR-8019 - Added audit event

### PR DESCRIPTION
**Pull request by Daveo-G**

### CODE REVIEW

File: app/uk/gov/hmrc/incorporatedentityidentificationfrontend/services/AuditService.scala [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+            }
+          val businessVerificationStatusBlock =
+            optBusinessVerificationStatus match {
+              case Some(bvStatus) => Json.obj("VerificationStatus" -> bvStatus)
+              case _ => Json.obj("VerificationStatus" -> BusinessVerificationFail)
+            }
+          val registrationStatusBlock =
+            optRegistrationStatus match {
+              case Some(regStatus) => Json.obj("RegisterApiStatus" -> regStatus)
+              case _ => Json.obj("RegisterApiStatus" -> RegistrationNotCalled)
```
Comment (kabuumu) [inline app/uk/gov/hmrc/incorporatedentityidentificationfrontend/services/AuditService.scala:58]: Does RegistrationNotCalled not get stored in the backend? We shouldn't have to default it should we?

File: app/uk/gov/hmrc/incorporatedentityidentificationfrontend/services/JourneyService.scala [hunk 2]
```diff
@@ -41,4 +41,12 @@ class JourneyService @Inject()(journeyConnector: JourneyConnector,
...
         throw new NotFoundException(s"Journey config was not found for journey ID $journeyId")
     }
+
+  def getCallingService(journeyId: String, authInternalId: String): Future[Option[String]] =
+    journeyConfigRepository.findJourneyConfig(journeyId, authInternalId).map {
+      case Some(journeyConfig) =>
+        journeyConfig.pageConfig.optServiceName
+      case _ =>
+        throw new NotFoundException(s"Journey config was not found for journey ID $journeyId when trying to get calling service")
+    }
```
Comment (kabuumu) [inline app/uk/gov/hmrc/incorporatedentityidentificationfrontend/services/JourneyService.scala:51]: Do we need a separate method here or could we just use the journey config model?

File: it/uk/gov/hmrc/incorporatedentityidentificationfrontend/controllers/RegistrationControllerISpec.scala [hunk 3]
```diff
@@ -193,6 +199,9 @@ class RegistrationControllerISpec extends ComponentSpecHelper with AuthStub with
           stubRetrieveBusinessVerificationStatus(testJourneyId)(status = OK, body = Json.toJson(BusinessVerificationPass))
           stubRegisteredSocietyRegister(testCompanyNumber, testCtutr)(status = OK, body = testFailedRegistration)
           stubStoreRegistrationStatus(testJourneyId, testFailedRegistration)(status = OK)
+          stubRetrieveIdentifiersMatch(testJourneyId)(status = OK, body = true)
+          stubRetrieveRegistrationStatus(testJourneyId)(status = OK, body = Json.toJson(testFailedRegistration))
+
```
Comment (kabuumu) [inline it/uk/gov/hmrc/incorporatedentityidentificationfrontend/controllers/RegistrationControllerISpec.scala:204]: Is the auditing being verified here?

File: it/uk/gov/hmrc/incorporatedentityidentificationfrontend/controllers/CheckYourAnswersControllerISpec.scala [hunk 4]
```diff
@@ -48,12 +61,14 @@ class CheckYourAnswersControllerISpec extends ComponentSpecHelper
...
       ))
       stubAuth(OK, successfulAuthResponse(Some(testInternalId)))
+      stubAudit()
       stubRetrieveCompanyProfileFromBE(testJourneyId)(status = OK, body = Json.toJsObject(testCompanyProfile))
       stubRetrieveCtutr(testJourneyId)(status = OK, body = testCtutr)
 
       lazy val result: WSResponse = get(s"$baseUrl/$testJourneyId/check-your-answers-business")
 
       result.status mustBe OK
+      verifyAudit()
```
Comment (kabuumu) [inline it/uk/gov/hmrc/incorporatedentityidentificationfrontend/controllers/CheckYourAnswersControllerISpec.scala:71]: Could this verify actually verify the correct information is audited please?

File: it/uk/gov/hmrc/incorporatedentityidentificationfrontend/views/CheckYourAnswersViewTests.scala [hunk 5]
```diff
@@ -118,11 +120,15 @@ trait CheckYourAnswersViewTests {
                               (result: => WSResponse,
                                companyNumberStub: => StubMapping,
                                authStub: => StubMapping,
-                               insertJourneyConfig: => Future[WriteResult]): Unit = {
+                               insertJourneyConfig: => Future[WriteResult],
+                               auditStub: => StubMapping,
+                               retrieveCtutrStub: => StubMapping): Unit = {
```
Comment (kabuumu) [inline it/uk/gov/hmrc/incorporatedentityidentificationfrontend/views/CheckYourAnswersViewTests.scala:125]: These need refactoring as we have done on some of the other services - we'll have to get a ticket on that.

File: project/AppDependencies.scala [hunk 6]
```diff
@@ -8,7 +8,7 @@ object AppDependencies {
     "uk.gov.hmrc" %% "bootstrap-backend-play-27" % "4.3.0",
     "uk.gov.hmrc" %% "play-frontend-govuk" % "0.71.0-play-27",
     "uk.gov.hmrc" %% "simple-reactivemongo" % "8.0.0-play-27",
-    "uk.gov.hmrc" %% "auth-client" % "5.4.0-play-27",
+    "uk.gov.hmrc" %% "auth-client" % "5.7.0-play-27",
```
Comment (kabuumu) [inline project/AppDependencies.scala:11]: auth-client is included as part of bootstrap so shouldn't need to be added here. The latest version of bootstrap has auth-client 2.6.0 though.

**Review by kabuumu — CHANGES_REQUESTED** [review]

File: it/uk/gov/hmrc/incorporatedentityidentificationfrontend/controllers/CheckYourAnswersControllerISpec.scala (hunk 4 above)
Comment (r-melvin) [reply it/uk/gov/hmrc/incorporatedentityidentificationfrontend/controllers/CheckYourAnswersControllerISpec.scala:71]: I think this is done at unit test level

File: app/uk/gov/hmrc/incorporatedentityidentificationfrontend/services/AuditService.scala (hunk 1 above)
Comment (Daveo-G) [reply app/uk/gov/hmrc/incorporatedentityidentificationfrontend/services/AuditService.scala:58]: This is for the 'catch all' case statement in CheckYourAnswersController (line 110).  This redirects to the CTR Mismatch page without saving the BVstatus or Register status.  Spoke to Ali and he still wanted this auditing but with the defaults.  There's a comment in the story for this.

File: app/uk/gov/hmrc/incorporatedentityidentificationfrontend/services/JourneyService.scala (hunk 2 above)
Comment (Daveo-G) [reply app/uk/gov/hmrc/incorporatedentityidentificationfrontend/services/JourneyService.scala:51]: I though the same thing, rewrote it and forgot to remove this method!!  I'll get it removed.

File: it/uk/gov/hmrc/incorporatedentityidentificationfrontend/controllers/CheckYourAnswersControllerISpec.scala (hunk 4 above)
Comment (Daveo-G) [reply it/uk/gov/hmrc/incorporatedentityidentificationfrontend/controllers/CheckYourAnswersControllerISpec.scala:71]: Yes, can do.  I lifted this from Soletrader.  There are some unit tests that test what the audit service produces.

File: app/uk/gov/hmrc/incorporatedentityidentificationfrontend/services/JourneyService.scala (hunk 2 above)
Comment (Daveo-G) [reply app/uk/gov/hmrc/incorporatedentityidentificationfrontend/services/JourneyService.scala:51]: I've removed this method

File: it/uk/gov/hmrc/incorporatedentityidentificationfrontend/controllers/RegistrationControllerISpec.scala (hunk 3 above)
Comment (Daveo-G) [reply it/uk/gov/hmrc/incorporatedentityidentificationfrontend/controllers/RegistrationControllerISpec.scala:231]: I've fixed this with a new method that verifies the explicit audit sent

File: it/uk/gov/hmrc/incorporatedentityidentificationfrontend/controllers/CheckYourAnswersControllerISpec.scala (hunk 4 above)
Comment (Daveo-G) [reply it/uk/gov/hmrc/incorporatedentityidentificationfrontend/controllers/CheckYourAnswersControllerISpec.scala:72]: I've fixed this with a new method that verifies the explicit audit sent

**Pull request merged**